4주차 과제1 (To-do 앱 Redux 사용하여 리팩터링하기)

seong-dodo opened the pull request:
안녕하세요.
오늘 과제 시작이 늦어져서 현재 진행사항까지 올립니다.

- `Redux 전환 과정 1` -  initialState 변수 할당
- `Redux 전환 과정 2` -  updateTaskTitle 함수 만들기
- `Redux 전환 과정 3` -  addTask 함수 만들기
- `Redux 전환 과정 4` -  deleteTask 함수 만들기

지금부터 추가로 진행하여 내일은 마감시간 내 최대한 많이 해서 제출하겠습니다~

seong-dodo committed:
Redux 전환 Step1

- initialState 변수 할당

seong-dodo committed:
Redux 전환(Step2)

- updateTaskTitle 함수 만들기

seong-dodo committed:
Redux 전환(Step3)

- addTask 함수 만들기

seong-dodo committed:
Redux 전환(Step4)

- deleteTask 함수 만들기

Kihyun92 reviewed (commented):
빠른 PR 좋습니다 👍🏼 

강의를 보고 따라하는것도 도움이 되지만, 직접 해보는건 훨씬 도움이 되죠 ㅎㅎ

![image](https://user-images.githubusercontent.com/35797540/119355265-06571a80-bce0-11eb-8bf5-aa4295a43da6.png)

배웠던 것을 떠올려 보면서 과제 조건에 맞게 진행해보시면 좋을 것 같아요~ 😊

seong-dodo committed:
강의따라하기 : 리덕스의 이해

- 리덕스로 상태관리 : action, store 분리
- App.test 수정
- mocking 모듈화

seong-dodo committed:
강의따라하기 : 리덕스의 이해

- store, reducer 분리

seong-dodo committed:
clone 받은 초기화 상태 만들기

- 강의따라 리덕스로 상태 분리 이해 후 초기화!

seong-dodo committed:
reducer 테스트 코드 추가1

- updateTaskTitle 테스트 추가

seong-dodo committed:
reducer 기능 구현1

- updateTaskTitle 동작 코드 구현

seong-dodo committed:
reducer 테스트 코드 추가2

- addTask 테스트 추가

seong-dodo committed:
reducer 기능 구현2 & 테스트 코드 수정

- addTask 동작 코드 구현
- addTask 테스트 코드 수정

seong-dodo commented:
https://github.com/CodeSoom/react-week4-assignment-1/pull/61#pullrequestreview-666740861

안녕하세요. 
피드백 주신 내용 참고하여 과제 진행하였습니다. 

오늘 updateTaskTitle, addTask 테스트 코드를 먼저 추가 한 후 기능 구현을 해보았습니다.
addTask 테스트 추가 후 기능을 구현해보았는데 기능 구현 과정에서 테스트 코드가 잘못 작성됨을 알게 되었습니다. 
테스트 코드를 완벽하게 만들지 못해서 이런 문제가 발생하여 기능구현 과정에서 테스트 코드를 수정하면서 TDD로 시작은 했지만 TDD가 아닌 TDD가 되었네요... 😢 😢 

다음 기능 개발에는 꼭 TDD 싸이클을 지켜보도록 하겠습니다... 🔥 🔥  🔥

seong-dodo committed:
refactoring

- actionCreator 사용

seong-dodo committed:
conflict resolve

seong-dodo committed:
reducer.test conflict resolve

seong-dodo committed:
reducer 테스트 코드 추가3

- addTask 테스트 코드 추가
- task title 있을 경우 & 없을 경우

seong-dodo committed:
reducer 기능 구현3

- addTask 동작 코드 구현 추가 (task title 없을 경우)

seong-dodo committed:
reducer 테스트 코드 추가4

- deleteTask 테스트 코드 추가(task id 있을 경우 & 없을 경우)

seong-dodo committed:
reducer 기능 구현4

- deleteTask 동작 코드 구현

seong-dodo committed:
deleteTask 리팩토링 하기

- actionCreator 사용

seong-dodo committed:
Provider & react-redux hook & 리듀서 테스트 인덴트 수정

- Provider 작업
- react-redux hook : useDispatch와 useSelector 사용
- 리듀서 테스트 인덴트 수정

seong-dodo commented:
https://github.com/CodeSoom/react-week4-assignment-1/pull/61#issuecomment-847866326

감사합니다 😆 
강의를 보고 기억력을 되살려 차근차근 해보았는데 완성된 기능 위주가 아닌 정말 처음부터 테스트 코드를 먼저 작성하고 기능 구현을 해본다는 생각으로 하나씩 작게 진행하도록 더 많은 고민이 필요한 것 같네요..
당장 컴포넌트를 render 하는 것 부터 하나씩 `RED - GREEN - REACTORING` 사이클로  과제2에 적용해보겠습니다.

이번 과제1는 커버리지 100% 달성할 수 있도록 추가 작업하여 올리도록 하겠습니다.

seong-dodo committed:
reducer 테스트 코드 추가 작성

- no action type 추가 및 reducer 테스트 커버리지 100% 달성

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,40 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === 'updateTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    const { newId, taskTitle, tasks } = state;
+
+    if (!taskTitle) {
+      return state;
+    }
+
+    return {
+      ...state,
+      newId: newId + 1,
+      tasks: [...tasks, { id: newId, title: taskTitle }],
+      taskTitle: '',
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    const { tasks } = state;
+
+    return {
+      ...state,
+      tasks: tasks.filter((task) => task.id !== action.payload.id),
+    };
+  }
```
`if`를 사용해서 action의 type에 따라 처리해주셨군요!

`object literal lookup`을 사용해보는 것도 좋을 것 같습니다~

- https://ultimatecourses.com/blog/deprecating-the-switch-statement-for-object-literals

seong-dodo replied:
안녕하세요 😭  😭  😭 
이번주 리덕스 이해 및 TDD로 과제 해결하는게 늦어져 차근차근 이해 및 수련 하면서 주어진 요구사항부터 완료 시킨 후 피드백 반영하도록 하겠습니다 🔥 🔥 🔥

Kihyun92 replied:
넵 의논이 필요한 부분이 있다면 말씀주세요!

Kihyun92 reviewed (changes requested):
두번째 과제도 기다리겠습니다~😄

seong-dodo committed:
redecer 테스트 코드 추가 작성에 대한 기능 구현

- no action type 추가에 대한 기능 구현

seong-dodo committed:
reducer 테스트 코드 주석 제거

- 리듀서 테스트 코드 주석 제거

seong-dodo committed:
InputContainer 테스트 코드 추가

- InputContainer 테스트 코드 추가

seong-dodo committed:
InputContainer 컴포넌트 만들기(리팩토링), 테스트 코드 추가 수정

- InputContainer 컴포넌트 만들고 코드 리팩토링
- InputContainer 테스트 코드 추가 수정

seong-dodo committed:
ListContainer 테스트 코드 추가

- ListContainer 테스트 코드 추가

seong-dodo committed:
ListContainer 컴포넌트 만들기(리팩토링), App컴포넌트 정리

- ListContainer 컴포넌트 만들고 코드 리팩토링
- App 컴포넌트 정리

seong-dodo committed:
App 컴포넌트와 Page 컴포넌트 결합

- Page컴포넌트 삭제
- App컴포넌트와 Page컴포넌트 결합
- 프레젠테이션 컴포넌트, 컨터이너 컴포넌트로 분리 완료

seong-dodo committed:
InputContainer 테스트 커버리지 완성 시키기

- props로 받은 메서드에 대한 테스트 검증 추가

seong-dodo committed:
ListContainer 테스트 커버리지 완성 시키기&  과제 요구사항 완료

- props로 받은 메서드에 대한 테스트 검증 추가
- ESLint, 인수테스트, 테스트 커버리지 100% 완료

Kihyun92 commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,41 @@
+
+  useSelector.mockImplementation((selector) => selector({
+    taskTitle: 'New Title',
+  }));
```
```suggestion
beforeEach(() => {
  dispatch.mockClear();
  useDispatch.mockImplementation(() => dispatch);

  useSelector.mockImplementation((selector) => selector({
    taskTitle: 'New Title',
  }));
});

```
`beforeEach`함수를 사용해서 테스트가 실행될 때마다 항상 일정한 상태에서 시작할 수 있도록 해주는게 좋을 것 같아요.
그러지 않으면 테스트가 다른 테스트들에 영향을 미칠 수 있고, 그러면 테스트가 올바르게 동작하지 않거나 유지보수하기 힘들어질 거에요.

Kihyun92 commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,41 @@
+    taskTitle: 'New Title',
+  }));
+
+  it('handleClickAddTask가 호출되는지 확인하는 테스트입니다', () => {
```
이 테스트에 대해서 알려주는 것도 좋지만, 사용 설명서 관점에서 이 테스트를 통해 무엇을 드러내고 싶은지 생각해보셔도 좋을 것 같습니다.

예를들면 `추가 버튼을 누르면 테스크가 추가됩니다` 는 어떤가요?

seong-dodo committed:
피드백 반영

- object literal 적용
- beforeEach 적용
- 테스트 시나리오 네이밍 수정

seong-dodo commented:
안녕하세요~ 
제가 코드숨 `main` 브랜치로 PR을 올려 다시 수정하였고, 피드백도 반영하였습니다~~~
